### PR TITLE
camera hardcode fix

### LINKED ISSUE #73: Fix Camera Hardcode
Camera Position has been hardcoded here
https://github.com/sdslabs/Rootex/blob/master/rootex/framework/components/visual/camera_visual_component.cpp#L14

This shouldn't be so and relevant hardcoded values should be removed from vertex shaders as well.
Possible add naive camera controls.

### PR DESCRIPTION
**Pull request by angad-k**
Fixes #73 <br>
The pull request aims to solve the following issues :<br>
- The CameraVisualComponent had hardcoded values and this led to a inflexibility and inability to use it as a normal component
- This led to the position of the camera being fixed at one position. 
---
The Pull request solves this issue in the following manner : <br>
- It adds CameraVisualComponent in the enum class ComponentIDs and adds REGISTER_COMPONENT(CameraVisualComponent) in the entity factory. This enables us to add and remove CameraVisualComponent to an entity through the inspector.
- It defines methods like CreateDefault() and Create(json& componentData) to the CameraVisualComponent class.
- It defines a restoreCamera() method in the RenderSystem class for going back to the default camera which is called when the active camera is either destroyed or made inactive. Also, it adds another step in the setCamera method in the same class which sets the m_Active variable of the previous camera to false.
- It redefines an getView() method in the CameraVisualComponent class which updates the Position, Up, Direction by getting the position and rotation from the owner Entity’s TransformComponent before making the view matrix with them.
---
In conclusion, this PR provides CameraVisualComponent the ability to be added to any entity as a normal component and functionality for it to derive its position and rotation from the owner entity’s TransformComponent.

### CODE REVIEW

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 1]
```diff
@@ -51,10 +104,117 @@ void CameraVisualComponent::postRender()
...
+			up.z = 0;
+			up.w = 0;
+			up = inverse * up * m_Rotation;
+			m_Up.x = up.x;
+			m_Up.y = up.y;
+			m_Up.z = up.z;
+		}
+		else
+		{
+			m_Up = { 0.0f, 1.0f, 0.0f };
```
Comment (sin3point14) [inline rootex/framework/components/visual/camera_visual_component.cpp:150]: Better to create these special directions as macros in common.h

File: rootex/framework/components/visual/camera_visual_component.h [hunk 2]
```diff
@@ -25,13 +34,25 @@ class CameraVisualComponent : public VisualComponent
 	virtual bool isVisible() const override;
 	virtual void postRender() override;
 
+	void onRemove() override;
 	virtual void setViewTransform(const Matrix& view);
+	void setNotActive();
```
Comment (sin3point14) [inline rootex/framework/components/visual/camera_visual_component.h:39]: setNorActive -> setInactive ?

File: rootex/framework/systems/render_system.cpp [hunk 3]
```diff
@@ -102,9 +103,15 @@ void RenderSystem::popUIMatrix()
 
 void RenderSystem::setCamera(Ref<CameraVisualComponent> camera)
 {
+	m_Camera->setNotActive();
 	m_Camera = camera;
 }
 
+void RenderSystem::restoreCamera()
+{
+	m_Camera = Ref<CameraVisualComponent>(new CameraVisualComponent());
```
Comment (sin3point14) [inline rootex/framework/systems/render_system.cpp:112]: Why create a new camera?
when no camera is set as the active camera, set a "default camera"

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 4]
```diff
@@ -1,24 +1,77 @@
 #include "camera_visual_component.h"
+#include "systems/render_system.h"
+Component* CameraVisualComponent::Create(const JSON::json& componentData)
```
Comment (sin3point14) [inline rootex/framework/components/visual/camera_visual_component.cpp:3]: Do not load Position and rotation from serialized file, used setup() in TransformComponent, refer conversation on slack

File: rootex/framework/systems/render_system.h [hunk 5]
```diff
@@ -27,6 +27,8 @@ class RenderSystem : public System
 
 	void setCamera(Ref<CameraVisualComponent> camera);
```
Comment (sin3point14) [inline rootex/framework/systems/render_system.h:29]: remove newline

File: rootex/framework/systems/render_system.cpp [hunk 6]
```diff
@@ -58,6 +58,7 @@ void RenderSystem::render()
 	RenderingDevice::GetSingleton()->setPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
 	RenderingDevice::GetSingleton()->setRasterizerState();
 	RenderingDevice::GetSingleton()->setDepthStencilState();
+	RenderSystem::GetSingleton()->getCamera()->updatePosition();
```
Comment (sin3point14) [inline rootex/framework/systems/render_system.cpp:61]: no need of GetSingleton()

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 7]
```diff
@@ -51,10 +104,117 @@ void CameraVisualComponent::postRender()
 void CameraVisualComponent::setPosition(Vector3 position)
 {
 	m_Position = position;
-	m_ViewMatrix = Matrix::CreateLookAt(m_Position, m_Position + m_Direction, { 0.0f, 1.0f, 0.0f });
+	m_ViewMatrix = Matrix::CreateLookAt(m_Position, m_Position + m_Direction, m_Up);
+}
+
+void CameraVisualComponent::updatePosition()
```
Comment (sin3point14) [inline rootex/framework/components/visual/camera_visual_component.cpp:110]: there shouldn't be any need of this, use TransformComponent of m_Owner

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 8]
```diff
@@ -1,24 +1,77 @@
...
 
 CameraVisualComponent::~CameraVisualComponent()
 {
 }
 
+void CameraVisualComponent::onRemove()
+{
+	if (m_Active)
+	{
+		if (RenderSystem::GetSingleton() != nullptr)
```
Comment (sin3point14) [inline rootex/framework/components/visual/camera_visual_component.cpp:69]: this will never happen

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 9]
```diff
@@ -51,10 +104,117 @@ void CameraVisualComponent::postRender()
...
+			m_Active = false;
+			RenderSystem::GetSingleton()->restoreCamera();
+		}
+	}
+	else
+	{
+		if (ImGui::Button("Set Active"))
+		{
+			m_Active = true;
+			auto cameraPointer = m_Owner->getComponent<CameraVisualComponent>();
```
Comment (sin3point14) [inline rootex/framework/components/visual/camera_visual_component.cpp:215]: no auto allowed except in range based for loops

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 10]
```diff
@@ -1,24 +1,78 @@
 #include "camera_visual_component.h"
+#include "systems/render_system.h"
+#include "DIRECTXMath.h"
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/camera_visual_component.cpp:3]: I don't think is the correct name, and/or this is required

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 11]
```diff
@@ -1,24 +1,78 @@
...
+Component* CameraVisualComponent::Create(const JSON::json& componentData)
+{
+	CameraVisualComponent* cameraVisualComponent = new CameraVisualComponent(
+	    { componentData["aspectRatio"]["x"], componentData["aspectRatio"]["y"]});
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/camera_visual_component.cpp:7]: Press TAB once

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 12]
```diff
@@ -1,24 +1,78 @@
...
+	
+	return cameraVisualComponent;
+}
 
-#include "entity_factory.h"
+CameraVisualComponent::CameraVisualComponent(const Vector2& aspectRatio)
+    : VisualComponent(RenderPassMain, false)
+    , m_DebugCamera(false)
+    , m_ViewMatrix(Matrix::CreateLookAt({ 0.0f, 0.0f, 0.4f }, { 0.0f, 0.0f, 0.0f }, { 0.0f, 1.0f, 0.0f }))
+    , m_ProjectionMatrix(Matrix::CreatePerspective(1.0f, 1.0f*aspectRatio.y/aspectRatio.x, 0.5f, 100.0f))
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/camera_visual_component.cpp:24]: TAB

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 13]
```diff
@@ -1,24 +1,78 @@
...
 
 CameraVisualComponent::CameraVisualComponent()
     : VisualComponent(RenderPassMain, false)
     , m_DebugCamera(false)
-    , m_ViewMatrix(Matrix::CreateLookAt({ 0.0f, 0.0f, 4.0f }, { 0.0f, 0.0f, 0.0f }, { 0.0f, 1.0f, 0.0f }))
-    , m_ProjectionMatrix(Matrix::CreatePerspective(1.0f, 1.0f * 480.0f / 640.0f, 0.5f, 100.0f))
-    , m_Active(true)
-    , m_Target(nullptr)
+    , m_ViewMatrix(Matrix::CreateLookAt({ 0.0f, 0.0f, 0.4f }, {0.0f, 0.0f, 0.0f}, { 0.0f, 1.0f, 0.0f }))
+    , m_ProjectionMatrix(Matrix::CreatePerspective(1.0f, 1.0f*9.0f/16.0f, 0.5f, 100.0f))
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/camera_visual_component.cpp:36]: TAB

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 14]
```diff
@@ -1,24 +1,78 @@
...
 }
 
+void CameraVisualComponent::onRemove()
+{
+	if (m_Active)
+	{
+		RenderSystem::GetSingleton()->restoreCamera();
+	}
+}
+bool CameraVisualComponent::setup()
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/camera_visual_component.cpp:55]: Newline before this

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 15]
```diff
@@ -48,13 +102,73 @@ void CameraVisualComponent::postRender()
...
 {
-	m_Position = position;
-	m_ViewMatrix = Matrix::CreateLookAt(m_Position, m_Position + m_Direction, { 0.0f, 1.0f, 0.0f });
+	if (m_TransformComponent)
+	{
+		Vector3 position = m_TransformComponent->getPosition();
+		Matrix transform = m_TransformComponent->getAbsoluteTransform();
+		Quaternion rotation = m_TransformComponent->getRotation();
+		Vector3 direction = transform.Forward();
+		Vector3 up = transform.Up();
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/camera_visual_component.cpp:113]: Change these to references instead. Saves us of extra copies

File: rootex/framework/components/visual/camera_visual_component.cpp [hunk 16]
```diff
@@ -48,13 +102,73 @@ void CameraVisualComponent::postRender()
...
+#include "imgui_stdlib.h"
+void CameraVisualComponent::draw()
+{
+	ImGui::DragFloat2("##A", &m_AspectRatio.x, s_EditorDecimalSpeed);
+	ImGui::SameLine();
+	if (ImGui::Button("Aspect Ratio"))
+	{
+		m_AspectRatio = { 16.0f, 9.0f };
+	}
+	m_ProjectionMatrix = Matrix::CreatePerspective(1.0f, 1.0f * m_AspectRatio.y / m_AspectRatio.x, 0.5f, 100.0f);
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/camera_visual_component.cpp:155]: Don't do this every frame. Do this only when DragFloat2 returns true i.e. it is edited by the user

File: rootex/framework/components/visual/camera_visual_component.h [hunk 17]
```diff
@@ -10,28 +10,46 @@ class CameraVisualComponent : public VisualComponent
...
 	virtual void render() override;
 	virtual bool reset(HierarchyGraph* visualComponentGraph, int windowWidth, int windowHeight);
 	virtual bool isVisible() const override;
 	virtual void postRender() override;
 
+	void onRemove() override;
 	virtual void setViewTransform(const Matrix& view);
-	void setPosition(Vector3 position);
+	void setInactive();
+	void updatePosition();
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/camera_visual_component.h:38]: Is this function not defined anywhere?

File: rootex/framework/systems/render_system.cpp [hunk 18]
```diff
@@ -12,6 +12,7 @@ RenderSystem* RenderSystem::GetSingleton()
 RenderSystem::RenderSystem()
     : m_Renderer(new Renderer())
     , m_Camera(new CameraVisualComponent())
+    , m_DefaultCamera(m_Camera)
```
Comment (IronicallySerious) [inline rootex/framework/systems/render_system.cpp:15]: Initialization in initialization lists happens in the opposite order of the way member data is specified in the class definition. This way, m_DefaultCamera will get nullptr as its value

**Comment by IronicallySerious** [comment]
Last round of changes ;)
